### DISCUSSION TITLE
Help with trusted header SSO and Roundcube via Nginx auth-request

### DISCUSSION THREAD

**Discussion by sphakka** [authelia/authelia#6064] [Troubleshooting] [answered]
I'm trying to set up authelia to authenticate Roundcube users via trusted headers. My testbed is a single bare-metal host with Nginx as reverse proxy (auth-request method). Please, bear with me as I'm learning how all this is supposed to work and might misunderstand things.

Roundcube comes with the [http_authenticatiion plugin](https://github.com/roundcube/roundcubemail/tree/master/plugins/http_authentication) which can be easily patched to use more standard headers, typically `HTTP_REMOTE_USER` instead of its `PHP_AUTH_USER`; then, Nginx is instructed to pass those headers as in [`authelia_proxy.conf`](https://www.authelia.com/integration/proxies/nginx/#authelia-authrequestconf):

    proxy_set_header Remote-User $user; 
    ...

The workflow (all happens on the same bare-metal installation) goes well until it hits the [Roundcube HTTP login routine](https://github.com/roundcube/roundcubemail/blob/master/plugins/http_authentication/http_authentication.php) which expects a user password coming from the header set as `PHP_AUTH_PW`. Since that's missing, the login fails.  As terrific as it may sounds (SSL is enabled and all internal traffic goes through localhost), is there a way to... ahem, expose the user password via a HTTP header? Cf. Lemonldap-ng the [allows doing it with all the related caveats](https://lemonldap-ng.org/documentation/2.0/passwordstore.html)...

**Comment by james-d-elliott** [comment, marked as answer]
Sorry to not get back to you sooner I've been rather busy. I had been meaning to respond to this earlier.

The complication with systems like roundcube and SMTP is unless they support OAuth 2.0 then it's fairly unlikely a feature will be able to be developed without badly impacting security, as you have noted it also requires passing the password to the backend in clear text. The passwords would have to be stored in an unhashed (effective plain-text) state for this to work, and Authelia would have to store them for at least the life of a session.

While the experience of it may be helpful to users the negatives to security (i.e. passwords stored in a recoverable format), and the potential for these values to be leaked to services where they shouldn't be makes the development and maintenance burden probably more than we want to take on.

Similar to other issues related to this we're unlikely to consider it, however I'll keep it in the back of my mind if I can figure out a sane way to do this without mindlessly enabling it for all users.

**Reply by sphakka** [reply]
Thanks for your reply, @james-d-elliott. Indeed, I suspected that this sort of "retrofitting" would incur the risk of compromising security to unacceptable levels. I guess LemonLDAP had to do that because of the legacy context where it was initially developed.

Meanwhile, I managed to integrate Roundcube + Dovecot with Authelia via OAuth  and will contribute soon some documentation :-)

**Reply by james-d-elliott** [reply]
We'd really welcome that. Evidently OAuth2.0 is a great mechanism for this due to the short lived authorization. In addition to the authorization itself there is the potential for backends to require 2FA and the OAuth2.0 provider is able to communicate that technology was used. While we don't have direct support for this yet (we only indicate the 2FA, we don't yet support clients advertising they want it) there are a lot more possibilities.. all without exchanging long-lived credentials like passwords.

Since many mail backends (dovecot and by extension postfix and many other SMTP clients) support XOAUTH it's not really a question what the best choice is. But yes, we really don't want to introduce features which could break fundamental security barriers.

**Comment by crabdancing** [comment]
Is using Authelia-controlled LDAP not a good idea for this use case for some reason?
